Update ssl configuration values

drming opened the pull request:
Update ssl configuration values to enable https

drming committed:
update ssl configuration

set ssl values if TLS enable

drming committed:
Merge branch 'master' of https://github.com/drming/pulsar-ansible

drming committed:
Merge branch 'datastax:master' into master

drming committed:
Adminconsole 2.x ssl

update for admin console 2.x ssl configuration and startup

yabinmeng commented on roles/adminconsole/ac_configAc/tasks/main.yaml:
```diff
@@ -89,11 +94,11 @@
-    - { regexp: "{{ placeHolder_sslKeyPath }}", line: "{{ ac_ssl_key_path }}" }
+    - { regexp: "{{ placeHolder_sslVerifyCerts }}", line: "{% if enable_brkr_tls is defined and enable_brkr_tls|bool %}false{% else %}true{% endif %}" }
+    - { regexp: "{{ placeHolder_sslHostnameValidation }}", line: "{% if enable_brkr_tls is defined and enable_brkr_tls|bool %}false{% else %}true{% endif %}" }
+    - { regexp: "{{ placeHolder_sslCaPath }}", line: "{% if enable_brkr_tls is defined and enable_brkr_tls|bool %}{{ tgt_adminConsole_sec_dir }}/tls/rootca/{{ public_cacert_name }}{% else %}{% endif %}" }
```
useless {% else %} block

drming replied:
Regarding "useless {% else %} block", first, the method is copied from original placeholder  "clientToken", and "caCertificate".  Second, if the double quote is in place of the filed in the local.json, then it will be "" and it will clear the value in the field in default.json.  if the no double quote is in place of the filed in the local.json, then it will be  nothing and it will not clear the value in the field in default.json.

yabinmeng replied:
ok

yabinmeng commented on roles/adminconsole/ac_configAc/tasks/main.yaml:
```diff
@@ -89,11 +94,11 @@
+    - { regexp: "{{ placeHolder_sslVerifyCerts }}", line: "{% if enable_brkr_tls is defined and enable_brkr_tls|bool %}false{% else %}true{% endif %}" }
+    - { regexp: "{{ placeHolder_sslHostnameValidation }}", line: "{% if enable_brkr_tls is defined and enable_brkr_tls|bool %}false{% else %}true{% endif %}" }
+    - { regexp: "{{ placeHolder_sslCaPath }}", line: "{% if enable_brkr_tls is defined and enable_brkr_tls|bool %}{{ tgt_adminConsole_sec_dir }}/tls/rootca/{{ public_cacert_name }}{% else %}{% endif %}" }
+    - { regexp: "{{ placeHolder_sslCertPath }}", line: "{% if enable_brkr_tls is defined and enable_brkr_tls|bool %}{{ tgt_adminConsole_sec_dir }}/tls/server/broker.{{ (hostvars[groups['broker'][0]]['private_ip'])|trim|replace('.','-') }}.crt.pem{% else %}{% endif %}" }
```
useless {% else %} block

drming replied:
see the reply above.

yabinmeng commented on roles/adminconsole/ac_configAc/tasks/main.yaml:
```diff
@@ -89,11 +94,11 @@
+    - { regexp: "{{ placeHolder_sslHostnameValidation }}", line: "{% if enable_brkr_tls is defined and enable_brkr_tls|bool %}false{% else %}true{% endif %}" }
+    - { regexp: "{{ placeHolder_sslCaPath }}", line: "{% if enable_brkr_tls is defined and enable_brkr_tls|bool %}{{ tgt_adminConsole_sec_dir }}/tls/rootca/{{ public_cacert_name }}{% else %}{% endif %}" }
+    - { regexp: "{{ placeHolder_sslCertPath }}", line: "{% if enable_brkr_tls is defined and enable_brkr_tls|bool %}{{ tgt_adminConsole_sec_dir }}/tls/server/broker.{{ (hostvars[groups['broker'][0]]['private_ip'])|trim|replace('.','-') }}.crt.pem{% else %}{% endif %}" }
+    - { regexp: "{{ placeHolder_sslKeyPath }}", line: "{% if enable_brkr_tls is defined and enable_brkr_tls|bool %}{{ tgt_adminConsole_sec_dir }}/tls/server/broker.{{ hostvars[groups['broker'][0]]['private_ip']|trim|replace('.','-') }}.key-pk8.pem{% else %}{% endif %}" }
```
useless {% else %} block

drming replied:
see the reply above.

yabinmeng commented on templates/ac_local.json.tmp:
```diff
@@ -12,8 +12,8 @@
         "ssl": {
-            "verify_certs": "<<ac_ssl_verify_certs>>",
-            "hostname_validation": "<<ac_ssl_hostname_validation>>",
+            "verify_certs": <<ac_ssl_verify_certs>>,
```
why removing double quote?

drming replied:
I tested many times and with double quotes, it will generate errors.  I checked the default.json and there is no double quotes for these two values.

drming replied:
"false" is not always the same as false. Most of developers will think they the same. false is boolean. "false" is a string. If a program detects the field with proper conversion build-in, "false" can be the same as false. Otherwise,  "false" is true in boolean. Beside my own testing, I noticed Mike ask me to remove the double quote in few case before we changed ssl values.

yabinmeng replied:
ok.

yabinmeng commented on roles/adminconsole/ac_configAc/tasks/main.yaml:
```diff
@@ -24,6 +24,7 @@
     recurse: yes
   with_items:
     - "{{ tgt_adminConsole_sec_dir }}/tls/rootca"
+    - "{{ tgt_adminConsole_sec_dir }}/tls/server"
```
Can you explain why broker's private key is needed?

The generated private key is unique and *private* to broker. It shouldn't be existing outside the broker. Otherwise, it is going to violate the security policy and the Pulsar cluster won't be secure any more.

AdminConsole, as a "client" connecting to the broker, I don't see why it needs broker's private key.

drming replied:
According to https://pulsar.apache.org/docs/security-tls-transport/, when the component is acting as proxy, these information will be needed to for establishing two way communication. So, I asked Chris B whether the new adminconsole as acting the proxy. His answer is "so like it". There is nothing wrong in security to configure two way communication with the same keys between two servers as long as they behind the firewalls and communication only only through limited port with limited protocol. You noticed that because you see the filename in the configuration file.  If you insist, we need to change it, we can change the filename

yabinmeng replied:
This is not correct. In any case, you can't share a broker's private key with any other component

yabinmeng replied:
It is not about changing the file name. One private key uniquely represents a server host. You can't share it among different server hosts. If AdminConsole needs its own private key, then you need to create its own private key.

drming replied:
I will change 01. script to add private key for adminconsole although I can not agree. The private key can not be shared only when private key contains both uuid type (hardware) information and dynamic (logical) target information (such as IP address). But, I will change the scripts as your request.

yabinmeng replied:
not sure which part you disagree. your last change is literally using the broker's private key on AdminConsole for HTTPS access purpose.

yabinmeng commented on roles/adminconsole/ac_configAc/tasks/main.yaml:
```diff
@@ -50,7 +51,11 @@
     group: "{{ pulsar_user_group }}"
     force: "{{ force_existing_adminConsole_tls_certs }}"
   with_items:
```
Same comment as before

yabinmeng commented on templates/ac_local.json.tmp:
```diff
@@ -12,8 +12,8 @@
-            "verify_certs": "<<ac_ssl_verify_certs>>",
-            "hostname_validation": "<<ac_ssl_hostname_validation>>",
+            "verify_certs": <<ac_ssl_verify_certs>>,
+            "hostname_validation": <<ac_ssl_hostname_validation>>,
```
why removing double quote?

yabinmeng reviewed (changes requested):
Please address the comments. Especially the first one of copying broker's private key to AdminConsole client machine.

drming committed:
Update admin console ssl key

Update admin console ssl key - create keys or adminconsole and apply the change in ssl configuration

yabinmeng commented on group_vars/all:
```diff
@@ -233,4 +233,25 @@ skip_fwrkr_tls_certs_generatation: false
+
+#
+# How to deploy Admin Console
+# - shared: deploy Admin Console as part of brokers
```
What does it mean by deploying AdminConsole as part of broker? Do you mean "AdminConsole" can run as part of the broker process?
Can you please elaborate this?

drming replied:
I updated the comment.

yabinmeng commented on group_vars/all:
```diff
@@ -233,4 +233,25 @@ skip_fwrkr_tls_certs_generatation: false
+deploy_adminconsole: "dedicated"
+
+# - Whether or not to enable Admin Console In-Transit TLS encryption
+enable_anco_tls: true
```
I would recommend using "enable_anco_https" as the variable name to make it clear that it is for HTTPS access.

drming replied:
Change it.

drming committed:
Add one line to force the commit

drming committed:
change parameter name

Change parameter name
